### PR TITLE
Both ends - show/hide panels (task #2322)

### PR DESCRIPTION
**Pull request by flangofas**
**What is the use case:**

When we want to conditionally render panels then we can specify in the config file the following:

In config:

**Structure:**
~~~~
[panels]
Panel Name A = "expression"
~~~~

**Example with values:**
~~~~
[panels]
Company Information = "(%%email%% == 'foobar@foobar.com' && %%location%% == 'other')"
Personal Information = "(%%email%% == 'foobar@foobar.com' || %%country%% == 'Cyprus')"
~~~~

**What is what:**
- Panel is the group of fields defined in the CsvMigrations views
- Expression is conditional expression which holds multiple conditions and their operators

The expression is evaluated using the [Expression Language component](http://symfony.com/doc/current/components/expression_language.html) which returns boolean value.
When the return is true, the panel is shown otherwise the panel is not rendered.

### CODE REVIEW

File: src/Controller/Component/CsvViewComponent.php [hunk 1]
```diff
@@ -94,6 +105,70 @@ public function beforeFilter(Event $event)
...
+            }
+        }
+    }
+
+    /**
+     * Parse the config file and set to the _parseConfig.
+     *
+     * @return void
+     */
+    protected function _parseConfig()
```
Comment (georgeconstantinou) [inline src/Controller/Component/CsvViewComponent.php]: @flangofas [ConfigurationTrait](https://github.com/QoboLtd/cakephp-csv-migrations/blob/master/src/ConfigurationTrait.php#L67-L74) already handles this and is called in base [Table](https://github.com/QoboLtd/cakephp-csv-migrations/blob/master/src/Table.php#L46) class.

Comment (flangofas) [inline src/Controller/Component/CsvViewComponent.php]: @georgeconstantinou ConfigurationTrait is conflicting with InstanceConfigTrait (used in Componet parent class)

File: src/Controller/Component/CsvViewComponent.php [hunk 2]
```diff
@@ -94,6 +106,137 @@ public function beforeFilter(Event $event)
...
+    {
+        //Read module's config
+        $this->_parseConfig();
+        if ($this->_isConditionalPanels()) {
+            $this->panelConditions();
+        }
+    }
+
+    /**
+     * From the given panels in the CsvMigration module config.
```
Comment (mamchenkov) [inline src/Controller/Component/CsvViewComponent.php]: @flangofas the recommended doc-block format is:

```
/**
 * Short description of the method (1 line)
 *
 * Long description, separated from short description by an
 * empty line.  Then, after another empty line, all the @-keywords
 * such as @throws, @return, @param, etc.
 *
 * @return void
 */
```

File: src/Controller/Component/CsvViewComponent.php (hunk 1 above)
Comment (georgeconstantinou) [inline src/Controller/Component/CsvViewComponent.php]: @flangofas you don't need to load ConfigurationTrait here. You can access table instance with `$this->_tableInstance` and you can fetch table's csv config with `$this->_tableInstance->getConfig()`

File: src/Controller/Component/CsvViewComponent.php [hunk 3]
```diff
@@ -94,6 +106,137 @@ public function beforeFilter(Event $event)
...
+     * Remove only the specified panels in the config which
+     * their type is not matching with the type field of the entity.
+     *
+     * @return bool|void
+     */
+    public function panelConditions()
```
Comment (mamchenkov) [inline src/Controller/Component/CsvViewComponent.php]: @flangofas method names that start with a verb (get, set, is, fetch, apply, etc) are much more readable.  They usually suggest what actions the method will perform and what is the likely return value.

File: src/Controller/Component/CsvViewComponent.php [hunk 4]
```diff
@@ -94,6 +106,137 @@ public function beforeFilter(Event $event)
...
+    {
+        $controller = $this->_registry->getController();
+        $fields = &$controller->viewVars['fields'];
+        $entity = $controller->viewVars['entity'];
+
+        $panels = $this->_parsedConfig[self::PANELS];
```
Comment (mamchenkov) [inline src/Controller/Component/CsvViewComponent.php]: @flangofas what happens if there is no panel configuration in the config file? As in, default behavior.

File: src/Controller/Component/CsvViewComponent.php (hunk 1 above)
Comment (flangofas) [inline src/Controller/Component/CsvViewComponent.php]: @georgeconstantinou I see.

File: src/Controller/Component/CsvViewComponent.php (hunk 2 above)
Comment (flangofas) [inline src/Controller/Component/CsvViewComponent.php]: @mamchenkov done!

File: src/Controller/Component/CsvViewComponent.php (hunk 3 above)
Comment (flangofas) [inline src/Controller/Component/CsvViewComponent.php]: @mamchenkov done.

File: src/Controller/Component/CsvViewComponent.php (hunk 4 above)
Comment (flangofas) [inline src/Controller/Component/CsvViewComponent.php]: @mamchenkov See https://github.com/QoboLtd/cakephp-csv-migrations/pull/129/files#diff-b0036d97b3452759d547aa252f152137R115

File: src/Controller/Component/CsvViewComponent.php (hunk 1 above)
Comment (flangofas) [inline src/Controller/Component/CsvViewComponent.php]: @georgeconstantinou done!

File: src/Controller/Component/CsvViewComponent.php [hunk 5]
```diff
@@ -94,6 +105,134 @@ public function beforeFilter(Event $event)
...
+    /**
+     * Compare first against the second value based on the given operator.
+     *
+     * @throws InvalidArgumentException Operator MUST be supported.
+     * @param  string $first    First value
+     * @param  string $operator Comparison operator
+     * @param  string $second   Second value
+     * @return bool
+     */
+    protected function _compare($first, $operator, $second)
```
Comment (mamchenkov) [inline src/Controller/Component/CsvViewComponent.php]: @flangofas have a look at this example:

```php
#!/usr/bin/php
<?php

$data = [ 
    'name' => 'Leonid',
    'surname' => 'Mamchenkov',
    'vip' => true,
    'year' => 1978,
];

$expression = '($name == "Leonid" || $year > 1970)';

print_r(evalCondition($expression, $data));

function evalCondition($expression, $data) {
    $result = null;
    
    $expression .= ' ? true : false';
    extract($data);
    eval("\$result = $expression;");
    return $result;
}
```

Obviously, we need to work on it a bit, because `eval()` is a dangerous thing altogether, but if we are doing evaluation of expressions, we need to find the suitable solution.  Either `eval()` or some form of expression parsing, like discussed in [this thread](http://stackoverflow.com/questions/327060/dynamic-logical-expression-parsing-evaluation-in-php).

Comment (flangofas) [inline src/Controller/Component/CsvViewComponent.php]: @mamchenkov makes sense. Will look into it.

Comment (flangofas) [inline src/Controller/Component/CsvViewComponent.php]: @mamchenkov From what I understand, this implementation need to support all logical operators as well. Correct?

Comment (mamchenkov) [inline src/Controller/Component/CsvViewComponent.php]: @flangofas the initial requirement was to support a single field and value condition.  You've started with the logic operators ... :)

**Comment by mamchenkov** [comment]
@flangofas overall, looks good.  But before I merge, I want some unit tests, so that we can test expressions and edge cases to see if the evaluation is done right.  Among other things, I want to see what happens when:

* Field values contain multiple words: `status = 'first attempt'`
* Field values contain special characters: `status = 'attempt #1'`
* Same keyword is used as field and value: `status = 'active' && active = false`

Once we have a unit test with expressions data provider, it'll be easy to throw in some other cases.

File: webroot/js/panels.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    };
+
+    Panel.prototype.observe = function() {
+        var that = this;
+        var $form = $('form');
+        var url = $form.attr('action');
+        var token = api_options.token;
+        $form.find(':input').change(function() {
+            $.ajax({
+                url: '/api/contacts/panels/',
```
Comment (mamchenkov) [inline webroot/js/panels.js]: @flangofas hardcoded URL for just contacts?

Comment (flangofas) [inline webroot/js/panels.js]: @mamchenkov Sending more commits.

Comment (flangofas) [inline webroot/js/panels.js]: @mamchenkov done.

File: src/Controller/Api/AppController.php [hunk 7]
```diff
@@ -210,6 +211,35 @@ protected function _prependParentModule(ResultSetDecorator $entities)
...
+        $this->request->allowMethod(['ajax', 'post']);
+        $result = [
+            'success' => false,
+            'data' => [],
+        ];
+        $table = $this->loadModel();
+        $tableConfig = $table->getConfig();
+        $entity = $table->newEntity($this->request->data);
+        $panels = Panel::getPanelNames($tableConfig) ?: [];
+        foreach ($panels as $name) {
```
Comment (mamchenkov) [inline src/Controller/Api/AppController.php]: @flangofas this functionality exists in more than one place ... here and in `beforeRender()` of the CsvViewComponent.php ... should be refactored into a method.

Comment (flangofas) [inline src/Controller/Api/AppController.php]: @mamchenkov done.

File: src/Controller/Api/AppController.php [hunk 8]
```diff
@@ -210,6 +212,33 @@ protected function _prependParentModule(ResultSetDecorator $entities)
...
+    {
+        $this->request->allowMethod(['ajax', 'post']);
+        $result = [
+            'success' => false,
+            'data' => [],
+        ];
+        $table = $this->loadModel();
+        $tableConfig = $table->getConfig();
+        $entity = $table->newEntity($this->request->data);
+        $panels = Panel::getPanelNames($tableConfig) ?: [];
```
Comment (mamchenkov) [inline src/Controller/Api/AppController.php]: @flangofas you don't seem to be using this variable anymore ...

Comment (flangofas) [inline src/Controller/Api/AppController.php]: @mamchenkov Done. Sry, missed it.

File: tests/TestCase/PanelTest.php [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        $panel->setName();
+    }
+
+    public function testGetExpression()
+    {
+        $panel = new Panel($this->first, $this->config);
+        $expression = $panel->getExpression();
+
+        //Should not be empty.
+        $this->assertTrue(!empty($expression));
```
Comment (mamchenkov) [inline tests/TestCase/PanelTest.php]: @flangofas `$this->assertFalse(empty($expression));` ... then no need for comment :)

File: tests/TestCase/PanelTest.php [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        $config['panels']['Foobar'] = "(%%type%% == 'foobar' && %%name%% == 'antonis')";
+        $panelB = new Panel($panelName, $config);
+
+        $this->assertTrue(in_array('type', $panelB->getFields()), 'Given field name is not in the array');
+        $this->assertTrue(in_array('name', $panelB->getFields()), 'Given field name is not in the array');
+    }
+
+    public function testEvalExpression()
+    {
+        $scenario = 'Scenario 1 - string comparison';
```
Comment (mamchenkov) [inline tests/TestCase/PanelTest.php]: @flangofas please use data provider for scenarios

File: tests/TestCase/PanelTest.php (hunk 9 above)
Comment (flangofas) [inline tests/TestCase/PanelTest.php]: @mamchenkov done.

File: tests/TestCase/PanelTest.php (hunk 10 above)
Comment (flangofas) [inline tests/TestCase/PanelTest.php]: @mamchenkov done.

File: tests/TestCase/PanelTest.php [hunk 11]
```diff
@@ -91,45 +88,26 @@ public function testGetFields()
...
-        $data = ['status' => 'attempt #1'];
-        $this->assertTrue($panel->evalExpression($data), sprintf('%s - Expression evaluation failed', $scenario));
-
-        $scenario = 'Scenario 3 - logical operator AND';
         $panelName = 'Foobar';
-        $config['panels']['Foobar'] = "%%status%% == 'active' && %%active%% == false";
+        $config['panels']['Foobar'] = $expression;
         $panel = new Panel($panelName, $config);
-        $data = ['status' => 'active', 'active' => false];
         $this->assertTrue($panel->evalExpression($data), sprintf('%s - Expression evaluation failed', $scenario));
```
Comment (mamchenkov) [inline tests/TestCase/PanelTest.php]: @flangofas it's better to provide the expected value from the data provider.  This way, things can be tested for false values, for example, as well.

Comment (flangofas) [inline tests/TestCase/PanelTest.php]: @mamchenkov done.
